mir_perception/mir_object_segmentation/common/src/scene_segmentation.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2018 Bonn-Rhein-Sieg University
+ *
+ * Author: Mohammad Wasil, Santosh Thoduka
+ *
+ */
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <mir_object_segmentation/scene_segmentation.h>
+
+SceneSegmentation::SceneSegmentation()
+{
+    cluster_extraction_.setSearchMethod(boost::make_shared<pcl::search::KdTree<PointT> >());
+    normal_estimation_.setSearchMethod(boost::make_shared<pcl::search::KdTree<PointT> >());
+
+    normal_estimation_omp_.setNumberOfThreads(4);
+    normal_estimation_omp_.setSearchMethod(boost::make_shared<pcl::search::KdTree<PointT> >());
+
+};
+SceneSegmentation::~SceneSegmentation()
+{
+
+};
+
+PointCloud::Ptr SceneSegmentation::segmentScene(const PointCloud::ConstPtr &cloud, std::vector<PointCloud::Ptr> &clusters,
+                                  std::vector<BoundingBox> &boxes, pcl::ModelCoefficients::Ptr &coefficients, double &workspace_height)
+{
+    PointCloud::Ptr filtered(new PointCloud);
+    PointCloud::Ptr plane(new PointCloud);
+    PointCloud::Ptr hull(new PointCloud);
+    pcl::PointIndices::Ptr segmented_cloud_inliers(new pcl::PointIndices);
+    std::vector<pcl::PointIndices> clusters_indices;
+
+    filtered = findPlane(cloud, hull, coefficients, workspace_height);
+
+    if (coefficients->values.size() == 0)
+    {
+        return  filtered;
+    }
+
+    extract_polygonal_prism_.setInputPlanarHull(hull);
+    extract_polygonal_prism_.setInputCloud(cloud);
+    extract_polygonal_prism_.setViewPoint(0.0, 0.0, 2.0);
+    extract_polygonal_prism_.segment(*segmented_cloud_inliers);
+
+    cluster_extraction_.setInputCloud(cloud);
+    cluster_extraction_.setIndices(segmented_cloud_inliers);
+    cluster_extraction_.extract(clusters_indices);
+
+    const Eigen::Vector3f normal(coefficients->values[0], coefficients->values[1], coefficients->values[2]);
+
+    for (size_t i = 0; i < clusters_indices.size(); i++)
+    {
+        const pcl::PointIndices& cluster_indices = clusters_indices[i];
+        PointCloud::Ptr cluster(new PointCloud);
+        pcl::copyPointCloud(*cloud, cluster_indices, *cluster);
+        clusters.push_back(cluster);
+        BoundingBox box = BoundingBox::create(cluster->points, normal);
+        boxes.push_back(box);
+    }
+    return filtered;
+}
+
+PointCloud::Ptr SceneSegmentation::findPlane(const PointCloud::ConstPtr &cloud, PointCloud::Ptr &hull,
+                              pcl::ModelCoefficients::Ptr &coefficients, double &workspace_height)
+{
+    PointCloud::Ptr filtered(new PointCloud);
+    PointCloud::Ptr plane(new PointCloud);
+    pcl::PointIndices::Ptr segmented_cloud_inliers(new pcl::PointIndices);
+
+    PointCloudN::Ptr normals(new PointCloudN);
+
+    voxel_grid_.setInputCloud(cloud);
+    voxel_grid_.filter(*filtered);
+
+    if (enable_passthrough_filter_)
+    {
+        pass_through_.setInputCloud(filtered);
+        pass_through_.filter(*filtered);
+    }
+
+    //normal_estimation_.setInputCloud(filtered);
```

Comment:
can be removed?

Reply:
I forgot this. Actually I did experiment with [OMP](http://pointclouds.org/documentation/tutorials/normal_estimation.php) to make it faster for normal estimation, but it didnt on my laptop. It may be faster on NUC since it has 8 cores.

Reply:
Normal estimation with OMP now is reconfigurable